x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java

```diff
@@ -0,0 +1,234 @@
+     */
+    private volatile boolean isElectedMaster;
+
+    public void initialize(ClusterService clusterService) {
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        isElectedMaster = event.localNodeMaster();
+        if (isElectedMaster == false) {
+            clear();
+        }
+    }
+
+    public void updateNodes(Client client, Set<DiscoveryNode> nodes, RerouteService rerouteService) {
+        if (isElectedMaster == false) {
+            return;
+        }
+
+        final List<AsyncNodeFetch> newFetches;
+        synchronized (mutex) {
+            // clean up nodes that left the cluster
+            nodeStates.keySet().removeIf(n -> nodes.contains(n) == false);
+
+            // skip nodes with known state
+            nodes.removeAll(nodeStates.keySet());
+
+            // reach out to any new nodes
+            newFetches = new ArrayList<>(nodes.size());
+            for (DiscoveryNode newNode : nodes) {
+                final NodeStateHolder nodeStateHolder = new NodeStateHolder();
+                final NodeStateHolder prevState = nodeStates.put(newNode, nodeStateHolder);
+                assert prevState == null;
+
+                logger.trace("fetching frozen cache state for {}", newNode);
+                newFetches.add(new AsyncNodeFetch(client, rerouteService, newNode, nodeStateHolder));
+            }
+        }
+
+        newFetches.forEach(Runnable::run);
+    }
+
+    public NodeState getNodeState(DiscoveryNode discoveryNode) {
+        final NodeStateHolder nodeStateHolder;
+        synchronized (mutex) {
+            nodeStateHolder = nodeStates.get(discoveryNode);
+        }
+        return nodeStateHolder == null ? NodeState.UNKNOWN : nodeStateHolder.nodeState;
+    }
+
+    public void clear() {
+        synchronized (mutex) {
+            nodeStates.clear();
+        }
+    }
+
+    public boolean isFetching() {
+        synchronized (mutex) {
+            return nodeStates.values().stream().anyMatch(nodeStateHolder -> nodeStateHolder.nodeState == NodeState.FETCHING);
+        }
+    }
+
+    private class AsyncNodeFetch extends AbstractRunnable {
+
+        private final Client client;
+        private final RerouteService rerouteService;
+        private final DiscoveryNode discoveryNode;
+        private final NodeStateHolder nodeStateHolder;
+
+        AsyncNodeFetch(Client client, RerouteService rerouteService, DiscoveryNode discoveryNode, NodeStateHolder nodeStateHolder) {
+            this.client = client;
+            this.rerouteService = rerouteService;
+            this.discoveryNode = discoveryNode;
+            this.nodeStateHolder = nodeStateHolder;
+        }
+
+        @Override
+        protected void doRun() {
+            client.admin()
+                .cluster()
+                .prepareNodesInfo(discoveryNode.getId())
```

Comment:
I'm wondering if this works with security (see #67416). As said before, I have a slight preference for a dedicated transport action.

Reply:
Ok, I'll work on an internal transport action for this next.

Reply:
Done in https://github.com/elastic/elasticsearch/pull/69904/commits/71b5bbb914b4fa644d1900e9f6584f290ba90739 now.